File: dashboard/src/lib/contracts/starter-kits/paygate-webhook-receiver.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"
	"time"
)

type webhookPayload struct {
	Transaction struct {
		OrderID string `json:"order_id"`
		Status  string `json:"status"`
	} `json:"transaction"`
}

func main() {
	secret := os.Getenv("PAYGATE_WEBHOOK_SECRET")
	if secret == "" {
		log.Fatal("PAYGATE_WEBHOOK_SECRET is required")
	}

	http.HandleFunc("/api/paygate/webhook", func(w http.ResponseWriter, r *http.Request) {
		timestamp := r.Header.Get("X-Webhook-Timestamp")
		signature := r.Header.Get("X-Webhook-Signature")
		if timestamp == "" || signature == "" {
			http.Error(w, `{"error":"missing webhook headers"}`, http.StatusBadRequest)
			return
		}

		value, err := strconv.ParseInt(timestamp, 10, 64)
		if err != nil || time.Since(time.Unix(value, 0)) > 5*time.Minute {
			http.Error(w, `{"error":"webhook timestamp is too old"}`, http.StatusBadRequest)
			return
		}

		rawBody, err := io.ReadAll(r.Body)
		if err != nil {
			http.Error(w, `{"error":"unable to read request body"}`, http.StatusBadRequest)
			return
		}

		mac := hmac.New(sha256.New, []byte(secret))
		mac.Write([]byte(timestamp))
		mac.Write([]byte("."))
		mac.Write(rawBody)
		expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))
		if !hmac.Equal([]byte(signature), []byte(expected)) {
			http.Error(w, `{"error":"invalid webhook signature"}`, http.StatusUnauthorized)
			return
		}

		var payload webhookPayload
		if err := json.Unmarshal(rawBody, &payload); err != nil {
			http.Error(w, `{"error":"invalid json body"}`, http.StatusBadRequest)
			return
		}

		log.Printf("PayGate webhook accepted order=%s status=%s\n", payload.Transaction.OrderID, payload.Transaction.Status)

		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, `{"received":true}`)
	})

	log.Println("PayGate webhook receiver listening on :3000")
	log.Fatal(http.ListenAndServe(":3000", nil))
}

File: dashboard/src/lib/contracts/starter-kits/paygate-webhook-receiver.php
<?php

declare(strict_types=1);

$secret = getenv('PAYGATE_WEBHOOK_SECRET');
$timestamp = $_SERVER['HTTP_X_WEBHOOK_TIMESTAMP'] ?? '';
$signature = $_SERVER['HTTP_X_WEBHOOK_SIGNATURE'] ?? '';
$rawBody = file_get_contents('php://input');

if (!$secret) {
    http_response_code(500);
    echo json_encode(['error' => 'PAYGATE_WEBHOOK_SECRET is required']);
    exit;
}

if ($timestamp === '' || $signature === '') {
    http_response_code(400);
    echo json_encode(['error' => 'Missing webhook headers']);
    exit;
}

$now = time();
if (abs($now - (int) $timestamp) > 300) {
    http_response_code(400);
    echo json_encode(['error' => 'Webhook timestamp is too old']);
    exit;
}

$expected = 'sha256=' . hash_hmac('sha256', $timestamp . '.' . $rawBody, $secret);
if (!hash_equals($expected, $signature)) {
    http_response_code(401);
    echo json_encode(['error' => 'Invalid webhook signature']);
    exit;
}

$payload = json_decode($rawBody, true, 512, JSON_THROW_ON_ERROR);

// Simpan perubahan status order ke database merchant.
error_log(sprintf(
    'PayGate webhook accepted for order %s with status %s',
    $payload['transaction']['order_id'],
    $payload['transaction']['status']
));

http_response_code(200);
header('Content-Type: application/json');
echo json_encode(['received' => true]);
